bundles/org.openhab.binding.philipsair/src/main/java/org/openhab/binding/philipsair/internal/connection/PhilipsAirAPIConnection.java

```diff
@@ -0,0 +1,242 @@
+
+    public PhilipsAirAPIConnection(PhilipsAirConfiguration config, HttpClient httpClient) {
+        this.httpClient = httpClient;
+        cache = new ExpiringCacheMap<>(TimeUnit.SECONDS.toMillis(config.getRefreshInterval()));
+        this.config = config;
+        initCipher();
+    }
+
+    private void initCipher() {
+        try {
+            this.cipher = new PhilipsAirCipher();
+            if (StringUtils.isEmpty(config.getKey())) {
+                exchangeKeys();
+            }
+
+            this.cipher.initKey(config.getKey());
+        } catch (GeneralSecurityException | ExecutionException | TimeoutException | InterruptedException e) {
+            logger.error("An exception occured", e);
+            this.cipher = null;
+        }
+    }
+
+    public synchronized @Nullable String getAirPurifierInfo(String host)
+            throws JsonSyntaxException, PhilipsAirAPIException {
+        return getResponseFromCache(buildURL(BASE_UPNP_URL, host));
+    }
+
+    public synchronized @Nullable PhilipsAirPurifierData getAirPurifierStatus(String host)
+            throws JsonSyntaxException, PhilipsAirAPIException {
+        return gson.fromJson(getResponseFromCache(buildURL(STATUS_URL, host)), PhilipsAirPurifierData.class);
+    }
+
+    public synchronized @Nullable String getAirPurifierKey(String host)
+            throws JsonSyntaxException, PhilipsAirAPIException {
+        return getResponseFromCache(buildURL(KEY_URL, host));
+    }
+
+    public synchronized @Nullable String getAirPurifierFirmware(String host)
+            throws JsonSyntaxException, PhilipsAirAPIException {
+        return getResponseFromCache(buildURL(FIRMWARE_URL, host));
+    }
+
+    public synchronized @Nullable PhilipsAirPurifierDevice getAirPurifierDevice(String host)
+            throws JsonSyntaxException, PhilipsAirAPIException {
+        return gson.fromJson(getResponseFromCache(buildURL(DEVICE_URL, host)), PhilipsAirPurifierDevice.class);
+    }
+
+    public synchronized @Nullable String getAirPurifierUserinfo(String host)
+            throws JsonSyntaxException, PhilipsAirAPIException {
+        return getResponseFromCache(buildURL(USERINFO_URL, host));
+    }
+
+    public synchronized @Nullable PhilipsAirPurifierFilters getAirPurifierFiltersStatus(String host)
+            throws JsonSyntaxException, PhilipsAirAPIException {
+        return gson.fromJson(getResponseFromCache(buildURL(FILTERS_URL, host)), PhilipsAirPurifierFilters.class);
+    }
+
+    private static String buildURL(String url, String host) {
+        return url.replaceFirst("%HOST%", host);
+    }
+
+    private @Nullable String getResponseFromCache(String url) {
+        return cache.putIfAbsentAndGet(url, () -> getResponse(url, HttpMethod.GET, null, true));
+    }
+
+    private String getResponse(String url, HttpMethod method, String content, boolean decode) {
+        try {
+            if (decode && cipher == null) {
+                logger.error("Cipher not initialized");

```

Comment:
The error logging level should be reserved for issues catastrophic enough to threaten the operation of openHab itself. An error in a binding shouldn't qualify for this. Please read the [openHab coding guidelines](https://www.openhab.org/docs/developer/guidelines.html#f-logging) for expected use of logging in bindings.

My general rule for logging levels is:
* trace - used for normal expected execution paths
* debug - used for unexepected execution paths, but well within normal operation
* info - used for notable points in an execution path, like a milestone. (in openhab we try reserve this logging level for the core, so bindings should rarely ever call this.)
* warn - used for notable unexpected execution paths that a regular user (not just a developer) should be notified of. Warnings should be used to indicate that something not-normal occurred and user intervention is required to resolve. Warnings do not indicate a failure to operate merely an abnormal condition of operation that can still be handled by the binding. Failures in binding operation should be indicated by changing the thing status to offline.
* error - used to indicate catastrophic program failure. This should be used to indicate a catastrophic failure in openhab's ability to operate. A failure in a binding would never cause openhab as a whole to fail so a failure in a bindings should never log an error. Instead that failure should be indicated by changing the thing status.

Reply:
True, I'm changing it to warn or info in some other cases